Add support for creating route and service by fetching image stream ports

rohitkrai03 opened the pull request:
This PR adds support for creating service and route resources based on available ports on the builder image.
- Fetches `ImageStreamTagModel` resource for the selected builder image and tag. Then takes out port data and saves it in the form.
- Also, fetched the same data when the user changes selected tag from the tag dropdown.
- Adds a `ExpandCollapse` component for advanced options.
- Adds a routing section in advanced section which only has a checkbox for creating a route to the application. It is set to `true` by default.

<img width="742" alt="Screenshot 2019-06-12 at 12 50 51 AM" src="https://user-images.githubusercontent.com/6041994/59300182-27b5cc00-8cac-11e9-8d47-f64daebc8d61.png">

rohitkrai03 committed:
Add support for creating route and service by fetching image stream ports

christianvogt commented on frontend/public/extend/devconsole/components/formik-fields/DropdownField.tsx:
```diff
@@ -19,10 +19,10 @@ const DropdownField: React.FC<DropdownFieldProps> = ({ label, helpText, ...props
         dropDownClassName={cx({ 'dropdown--full-width': props.fullWidth })}
         onChange={(value: string) => setFieldValue(props.name, value)}
         onBlur={() => setFieldTouched(props.name, true)}
+        {...props}
```
spreading props should always come first otherwise they can override the props this component sets. It should come even before `id`.

rohitkrai03 replied:
I know, but in this case i want the parent component to be able to override the `onChange` or `onBlur` prop. Is there any other way i can do this?

christianvogt replied:
Do you really want to override or just add the ability for them to also get the call back in addition to the default? If that's the case you need to manually callback in your own handlers.

rohitkrai03 replied:
I actually redid the whole logic of fetching ports using `React.useEffect`. Now i don't need to override `onChange` handlers and there is no duplicate logic in `BuilderImageSelector` and `BuilderImageTagSelector`.

christianvogt commented on frontend/public/extend/devconsole/components/import/GitImportForm.scss:
```diff
@@ -0,0 +1,5 @@
+.odc-git-import-form {
+  &__button-bar {
+    margin-top: var(--pf-global--spacer--md);
```
Nothing else that can be used? Maybe just a simple `<br />` otherwise each form needs to do the same thing.

rohitkrai03 replied:
Well, using `<br />` works fine but i was not sure if that is the correct way. I'll replace it with a `<br />`

christianvogt commented on frontend/public/extend/devconsole/components/import/builder/BuilderImageSelector.tsx:
```diff
@@ -23,18 +25,28 @@ const BuilderImageSelector: React.FC<BuilderImageSelectorProps> = ({
       },
     };
     setValues(newValues);
+    await k8sGet(ImageStreamTagModel, `${image}:${recentTag}`, imageStreamNamespace).then(
```
Would be good to cache these values.

rohitkrai03 replied:
Can we introduce a caching layer in `coFetch` for get requests like these? Not sure if it has some already.

christianvogt replied:
Ideally it's stored in redux. Firehose does this but also opens a socket. Think about it. We can open another pr to improve this.

christianvogt commented on frontend/public/extend/devconsole/components/import/builder/BuilderImageTagSelector.tsx:
```diff
@@ -27,23 +36,32 @@ const BuilderImageTagSelector: React.FC<BuilderImageTagSelectorProps> = ({
 
+  const handleImageTagChange = async (tag: string) => {
+    setFieldValue('image.tag', tag);
+    await k8sGet(ImageStreamTagModel, `${imageName}:${tag}`, imageStreamNamespace).then(
```
Same code as another section and again can we cache the result?

rohitkrai03 replied:
Yes, it's just fetching the data and setting something in the form. I tried moving it out to a utility function but still the handling of `.then` part of the Promise needs to be done here since it's using `setFieldValue` unless i pass that back to the function and not sure if that is a correct pattern though.

sbose78 commented:
build and deploy

sbose78 commented:
Your build is progress : https://console-openshift-console.apps.rgarg13.devcluster.openshift.com/k8s/ns/openshift-console/builds
 
 PR will be LIVE in https://pr-133-openshift-console.apps.rgarg13.devcluster.openshift.com

vikram-raj reviewed (commented):
Why ```FormSectionDivider``` style is different? At some places, it is darker.

![Screenshot from 2019-06-12 16-43-45](https://user-images.githubusercontent.com/2561818/59347030-490ccb80-8d31-11e9-80e9-1cc86cae2865.png)

See the divider above builderImage section is less dark than the divider after builderImage section.

vikram-raj reviewed (commented):
Run prettier. At some places trailing comma and space is missing.

rohitkrai03 commented:
> Why `FormSectionDivider` style is different? At some places, it is darker.
> 
> ![Screenshot from 2019-06-12 16-43-45](https://user-images.githubusercontent.com/2561818/59347030-490ccb80-8d31-11e9-80e9-1cc86cae2865.png)
> 
> See the divider above builderImage section is less dark than the divider after builderImage section.

Not sure why this is happening. Same styles are being applied in each section divider.

vikram-raj commented:
@rohitkrai03 Change the ```border``` to ```borderTop``` in ```FormSectionDivider```. This will fix the issue.

rohitkrai03 committed:
Move the port fetching logic to useEffect hook to simplify things

rohitkrai03 commented:
@christianvogt Made some changes to the PR. Changed fetching logic, fixed border issue with divider and also added logic for disabling create button when form has errors. Review again.
